### PR TITLE
Reworking critical technology concepts

### PR DESCRIPTION
**Pull request by FelixSchladt**
I propose to remove the concept folder completely after the concpets have been added to readthedocs.

Further I would like more concept papers:

* Testing framework
* Documentation aka why readthedocs?

### CODE REVIEW

File: concepts/ipc.md [hunk 1]
```diff
@@ -1,4 +1,8 @@
-# Concept IPC
+# IPC
+
+During a first discussion it was decided that oxide should be controllable via an ipc mechanism.
```
Comment (gungula) [inline concepts/ipc.md:3]:
```suggestion
During a first discussion it was decided that *Oxide* should be controllable via an ipc mechanism.
```

File: concepts/ipc.md [hunk 2]
```diff
@@ -7,13 +11,37 @@ This is neccessary for:
 * External libraries
 * Command line utility
 
-## Requirements
+## Feature List
```
Comment (gungula) [inline concepts/ipc.md:14]:
```suggestion
## Feature list
```

File: concepts/ipc.md [hunk 3]
```diff
@@ -7,13 +11,37 @@ This is neccessary for:
...
+The following list includes currently proposed features
+
+* Everything that is achievable via the keyboard (kill, move, launch...)
```
Comment (gungula) [inline concepts/ipc.md:17]:
```suggestion
* everything that is achievable via the keyboard (kill, move, launch...)
```

File: concepts/ipc.md [hunk 4]
```diff
@@ -7,13 +11,37 @@ This is neccessary for:
...
+* Current state of the window manager including layout, windows...
```
Comment (gungula) [inline concepts/ipc.md:18]:
```suggestion
* current state of the window manager including e.g. layout or windows
```

File: concepts/ipc.md [hunk 5]
```diff
@@ -7,13 +11,37 @@ This is neccessary for:
...
+
+
+## IPC Integration Solution
```
Comment (gungula) [inline concepts/ipc.md:21]:
```suggestion
## IPC integration solution
```

File: concepts/ipc.md [hunk 6]
```diff
@@ -7,13 +11,37 @@ This is neccessary for:
...
+
+Since there are two types of events that have to be handeled, there
```
Comment (gungula) [inline concepts/ipc.md:23]:
```suggestion
Since there are two types of events that have to be handled, there
```

File: concepts/ipc.md [hunk 7]
```diff
@@ -7,13 +11,37 @@ This is neccessary for:
...
+needs to be some separation bewteen them.
```
Comment (gungula) [inline concepts/ipc.md:24]:
```suggestion
needs to be some separation between them.
```

File: concepts/ipc.md [hunk 8]
```diff
@@ -7,13 +11,37 @@ This is neccessary for:
...
+
+One type are `xevents`, received from the `X11` instance, and the other one
+is custom events create by the user, recieved over `zbus`.
```
Comment (gungula) [inline concepts/ipc.md:27]:
```suggestion
is custom events created by the user, received over `zbus`.
```

File: concepts/ipc.md [hunk 9]
```diff
@@ -7,13 +11,37 @@ This is neccessary for:
...
+
+For this reason, each type of event will get its own loop on its own thread,
+which will await them and push them into a list shared between them.
+The events in this list will be taken care of by the window manager,
+who will execute the correct action based on event type and content.
+
+![IPC-QUEUE](ipc-queue.png)
+
+## Technical Solution
```
Comment (gungula) [inline concepts/ipc.md:36]:
```suggestion
## Technical solution
```

File: concepts/ipc.md [hunk 10]
```diff
@@ -7,13 +11,37 @@ This is neccessary for:
...
+The following sections describe the argument for the different ipc-mechanisms and libraries.
+
+### Requirements
 As for the aforementioned use cases it will not be required to send large amounts of data. 
 Only short messages will be exchanged between the clients. Also it is not expected that the ipc performance will have a significant impact on the usability of the system.
 Therefore some ipc options such as shared memory and semaphores will not be regarded as these options are not as easy to use and do not offer any significant advantages.
 
-## Research
-### Options
+### Possible IPC Mechanisms
```
Comment (gungula) [inline concepts/ipc.md:44]:
```suggestion
### Possible IPC mechanisms
```

File: concepts/ipc.md [hunk 11]
```diff
@@ -53,14 +81,14 @@ There are multiple different ways of implementing ipc on posix systems.
 * D-Bus should offer a high level, easy to use ipc mechanism
```
Comment (gungula) [inline concepts/ipc.md:81]: The word "easy" must not be contained in any of our documents. Please think about another formulation :)

File: concepts/ipc.md [hunk 12]
```diff
@@ -53,14 +81,14 @@ There are multiple different ways of implementing ipc on posix systems.
...
+### Conclusion
 
 After a technical discussion with our team, we came to the conclusion that **D-Bus is most suitable**.
 The **performance is deemed non critical** in our use case and the ease of use will be benefitial for the project.
 None the less, the ipc interface should be **created in an abstract manner** allowing for a possible replacement of the underlying ipc mechanism.  
 
-## Implementation
-### Available Libraries
+### Implementation
+#### Available Libraries
```
Comment (gungula) [inline concepts/ipc.md:91]:
```suggestion
#### Available libraries
```

File: concepts/ipc.md [hunk 13]
```diff
@@ -76,7 +104,7 @@ There seem to be two main projects striving to provide dbus support for rust.
 * Wrapper library for libdbus -> libdbus dependency
 * Examples
 
-### Conclusion
+#### Conclusion
 Zbus seems to have some advantages over dbus-rs, mainly:
 * Official freedesktop.org library
 * Pure rust -> no ibdbus dependency
```
Comment (gungula) [inline concepts/ipc.md:110]:
```suggestion
* pure rust -> no ibdbus dependency
```

File: concepts/ipc.md [hunk 14]
```diff
@@ -76,7 +104,7 @@ There seem to be two main projects striving to provide dbus support for rust.
 * Wrapper library for libdbus -> libdbus dependency
 * Examples
 
-### Conclusion
+#### Conclusion
 Zbus seems to have some advantages over dbus-rs, mainly:
 * Official freedesktop.org library
```
Comment (gungula) [inline concepts/ipc.md:109]:
```suggestion
* official freedesktop.org library
```

File: concepts/ipc.md (hunk 13 above)
Comment (gungula) [inline concepts/ipc.md:110]: Is the term "ibdbus" correct? I don't know it and i am not sure if it is a typo or not

File: concepts/reactiveness.md [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+In order to assure this polling of threads should never be implemented.
+
+## Event Sources
+
+Events for the window manager can be created by multiple sources. 
+
+1. `X-Events` Generated by the X-Server and regarding userinput, clientsrequests and more.
+2. `IPC-Events` Generated via the ipc-mechanism by the user or programs running on the system.
+
+In order to be able to handle these two event sources two possible solution are available
```
Comment (gungula) [inline concepts/reactiveness.md:13]:
```suggestion
In order to be able to handle these two event sources two possible solution are available:
```

File: concepts/reactiveness.md [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+
+
+1. Polling of both event-sources
+
+-->--|
+|    V
+|	Poll for X-Event
```
Comment (gungula) [inline concepts/reactiveness.md:20]:
```suggestion
|	poll for X-Event
```

File: concepts/reactiveness.md [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+|	 |
+|	 V
+|	Poll for ipc-event
```
Comment (gungula) [inline concepts/reactiveness.md:23]:
```suggestion
|	poll for ipc-event
```

File: concepts/reactiveness.md [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+|	 |
+|	 V
+| 	 Execute received event
```
Comment (gungula) [inline concepts/reactiveness.md:26]:
```suggestion
| 	 execute received event
```

File: concepts/reactiveness.md [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+--<---
+
+
+This layout allows for a very simple single threaded implementation of the window manager but requires a frequen polling loop in order to allow for a reactive behaviour.
```
Comment (gungula) [inline concepts/reactiveness.md:30]:
```suggestion
This layout allows for a very simple single threaded implementation of the window manager but requires a frequent polling loop in order to allow for a reactive behavior.
```

File: concepts/reactiveness.md [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+
+Therefore this layout is not suitable to our requirements as this would be very inefficient.
+
+2. Multithreaded Queues
```
Comment (gungula) [inline concepts/reactiveness.md:34]:
```suggestion
2. Multithreaded queues
```

File: concepts/reactiveness.md [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+
+![Queue Layout](ipc_queues.png)
+
+This layout requires three different queues.
+1. X-Event queue -> waits until an x-event is received
+2. IPC-Event queu -> waits until an ipc-event is reveived
```
Comment (gungula) [inline concepts/reactiveness.md:40]:
```suggestion
2. IPC-Event queue -> waits until an ipc-event is reveived
```

File: concepts/reactiveness.md [hunk 22]
```diff
@@ -0,0 +1,64 @@
...
+3. Mainloop -> waits for an event forwarded to the mainevent-queue by the above mentioned threads
```
Comment (gungula) [inline concepts/reactiveness.md:41]:
```suggestion
3. mainloop -> waits for an event forwarded to the mainevent-queue by the above mentioned threads
```

File: concepts/reactiveness.md [hunk 23]
```diff
@@ -0,0 +1,64 @@
...
+
+If implemented in the above shown way, no polling is required. This therefore allows to sleep the threads until an event is received and the kernel wakes up the thread. 
+It therefore will be very reactive and use only minimal resources.
+
+**=> Therfore we will implement this layout**
```
Comment (gungula) [inline concepts/reactiveness.md:46]:
```suggestion
**=> Therfore we will implement this layout.**
```

File: concepts/reactiveness.md [hunk 24]
```diff
@@ -0,0 +1,64 @@
...
+
+## Technical Solution
```
Comment (gungula) [inline concepts/reactiveness.md:48]:
```suggestion
## Technical solution
```

File: concepts/reactiveness.md [hunk 25]
```diff
@@ -0,0 +1,64 @@
...
+
+### Waiting for Event
```
Comment (gungula) [inline concepts/reactiveness.md:50]:
```suggestion
### Waiting for event
```

File: concepts/reactiveness.md [hunk 26]
```diff
@@ -0,0 +1,64 @@
...
+
+1. X-Event
+The `x11rb` library supports waiting on an event:
+[wait_for_event method](https://smithay.github.io/smithay/x11rb/connection/trait.Connection.html#method.wait_for_event)
+
+2. IPC-Event
+Zbus is asynchronous in nature. The handling functions are registered to the server and executed when an event is received.
+[Zbus Example](https://docs.rs/zbus/latest/zbus/)
+
+3. Mainloop
```
Comment (gungula) [inline concepts/reactiveness.md:61]:
```suggestion
3. mainloop
```

File: concepts/rust.md [hunk 27]
```diff
@@ -0,0 +1,15 @@
+# Rust as Implementation Language
```
Comment (gungula) [inline concepts/rust.md:1]:
```suggestion
# Rust as implementation language
```

File: concepts/rust.md [hunk 28]
```diff
@@ -0,0 +1,15 @@
...
+This document outlines why rust has been chosen as the language of implementation for this project.
+But it needs to be mentioned that this choice was not purely made out of technical reasoning, because we think rust is fun and we enjoy to learn this language.
+
+## Technical Arguments for Rust
+
+Rust is a good choice for a system level language as it allows direct memory acces but also object orientation and offers a memory safe environment.
+Further this is achieved without a resource intensive garbage collection and high speeds in the realm of C are possible.
+Even though rust is still "relatively" new, it already has a stable ecosystem and a lot of libraries supporting it.
+
+A first search for rust libraries allowing connections to the X-Server shows multiple results. The same is true for DBUS and IPC-Libraries.
```
Comment (gungula) [inline concepts/rust.md:12]:
```suggestion
A first search for rust libraries allowing connections to the X-Server shows multiple results. The same is true for dbus- and IPC-libraries. 
```

File: concepts/rust.md [hunk 29]
```diff
@@ -0,0 +1,15 @@
+# Rust as Implementation Language
+
+This document outlines why rust has been chosen as the language of implementation for this project.
+But it needs to be mentioned that this choice was not purely made out of technical reasoning, because we think rust is fun and we enjoy to learn this language.
+
+## Technical Arguments for Rust
```
Comment (gungula) [inline concepts/rust.md:6]:
```suggestion
## Technical arguments for Rust
```

**Review by gungula — CHANGES_REQUESTED** [review]

File: concepts/ipc.md (hunk 11 above)
Comment (B1TC0R3) [reply concepts/ipc.md:81]: "simple"

**Review by gungula — APPROVED** [review]

**Pull request merged**